### PR TITLE
The stmts property may exist but may be null

### PR DESCRIPTION
**Pull request by spaze**
- [x] I've read the disclaimer "No support is provided outside of bugs impacting development and usage of PHPStan" 😅

I'm trying to use this tool so I can develop my PHPStan extension with modern PHP constructs and then downgrading it so that it can support older PHPs too. I'm not yet sure how exactly it will work, if at all.

Here's one issue I've hit. The `stmts` property can be null in `PhpParser\Node\Stmt\ClassMethod` of abstract methods:
https://github.com/nikic/PHP-Parser/blob/1bcbb2179f97633e98bbbc87044ee2611c7d7999/lib/PhpParser/Node/Stmt/ClassMethod.php#L20-L21

### CODE REVIEW

**Comment by ondrejmirtes** [comment]
Please try to write failing test first, it's a mystery to me no existing code crashed on this :)

**Comment by spaze** [comment]
This is the minimal file the downgrader is crashing on:
```
<?php
declare(strict_types = 1);

class Foo
{

	abstract public function bar();

	final public function __construct(
		private readonly int $baz,
	) {
	}

}
```

```
$ vendor/bin/simple-downgrade downgrade --configuration build/downgrade.php 7.2
PHP Fatal error:  Uncaught TypeError: count(): Argument #1 ($value) must be of type Countable|array, null given in .../vendor/ondrejmirtes/simple-downgrader/src/Php/PhpPrinterIndentationDetectorVisitor.php:42Stack trace:
#0 .../vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(200): SimpleDowngrader\Php\PhpPrinterIndentationDetectorVisitor->enterNode()
#1 .../vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(114): PhpParser\NodeTraverser->traverseArray()
#2 .../vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(223): PhpParser\NodeTraverser->traverseNode()
#3 .../vendor/nikic/php-parser/lib/PhpParser/NodeTraverser.php(91): PhpParser\NodeTraverser->traverseArray()
#4 .../vendor/ondrejmirtes/simple-downgrader/src/Php/PhpPrinter.php(41): PhpParser\NodeTraverser->traverse()
#5 .../vendor/nikic/php-parser/lib/PhpParser/PrettyPrinterAbstract.php(491): SimpleDowngrader\Php\PhpPrinter->preprocessNodes()
#6 .../vendor/ondrejmirtes/simple-downgrader/src/Console/DowngradeCommand.php(168): PhpParser\PrettyPrinterAbstract->printFormatPreserving()
#7 .../vendor/ondrejmirtes/simple-downgrader/src/Console/DowngradeCommand.php(133): SimpleDowngrader\Console\DowngradeCommand->processFile()
#8 .../vendor/symfony/console/Command/Command.php(298): SimpleDowngrader\Console\DowngradeCommand->execute()
#9 .../vendor/symfony/console/Application.php(1040): Symfony\Component\Console\Command\Command->run()
#10 .../vendor/symfony/console/Application.php(301): Symfony\Component\Console\Application->doRunCommand()
#11 .../vendor/symfony/console/Application.php(171): Symfony\Component\Console\Application->doRun()
#12 .../vendor/ondrejmirtes/simple-downgrader/bin/simple-downgrade(39): Symfony\Component\Console\Application->run()
#13 .../vendor/bin/simple-downgrade(119): include('...')
#14 {main}
  thrown in .../vendor/ondrejmirtes/simple-downgrader/src/Php/PhpPrinterIndentationDetectorVisitor.php on line 42
```

I'll see if I can write a testcase, failing so far 😅

**Comment by ondrejmirtes** [comment]
Should be possible here https://github.com/ondrejmirtes/simple-downgrader/blob/1.x/tests/Visitor/DowngradeReadonlyPromotedPropertyVisitorTest.php otherwise I don't know how can it work :)

**Comment by spaze** [comment]
Tried creating a separate test case for `PhpPrinterIndentationDetectorVisitor` but then switched to updating `DowngradeReadonlyPromotedPropertyVisitorTest` as you suggest.

But I'm observing a different behavior in the test, and when I'm trying to run `simple-downgrade`. I don't know why, if you have a sec, can you please check or give a hint?

I have pushed the test and the file (both, minimized), the visitor has a `var_dump()` and `var_dump()` there.
https://github.com/spaze/simple-downgrader/tree/bug-3
If you clone the repo, switch to the `bug-3` branch and then run
```
bin/simple-downgrade downgrade --configuration build/downgrade.php 7.2
```
and 
```
vendor/bin/phpunit tests/Visitor/DowngradeReadonlyPromotedPropertyVisitorTest.php
```
you should see this output when run the tool
```
string(28) "PhpParser\Node\Stmt\Declare_"
string(34) "PhpParser\Node\Stmt\DeclareDeclare"
string(25) "PhpParser\Node\Identifier"
string(29) "PhpParser\Node\Scalar\LNumber"
string(26) "PhpParser\Node\Stmt\Class_"
string(28) "PhpParser\Node\Stmt\Property" <---------- this is not here when running the test
string(31) "PhpParser\Node\Stmt\ClassMethod"
string(31) "PhpParser\Node\Stmt\ClassMethod"
int(0)                                                       <------- the regexp matches nothing
string(0) ""
string(25) "PhpParser\Node\Identifier"
string(28) "PhpParser\Node\Stmt\Property"
string(36) "PhpParser\Node\Stmt\PropertyProperty"
string(32) "PhpParser\Node\VarLikeIdentifier"
string(31) "PhpParser\Node\Stmt\ClassMethod"     <------------------- this is the abstract method with no stmts
PHP Fatal error:  Uncaught TypeError: count(): Argument #1
```
The `<------` markers are what seems the be the reason why it crashes, but when you run the test, it behaves differently:
```
string(28) "PhpParser\Node\Stmt\Declare_"
string(34) "PhpParser\Node\Stmt\DeclareDeclare"
string(25) "PhpParser\Node\Identifier"
string(29) "PhpParser\Node\Scalar\LNumber"
string(26) "PhpParser\Node\Stmt\Class_"
string(31) "PhpParser\Node\Stmt\ClassMethod"
string(31) "PhpParser\Node\Stmt\ClassMethod"
int(3)
string(14) "class Foo
{

        "
"

I can't get the test to crash the same way as when running the tool. I'm most probably overlooking something but I'd appreciate your help here.

**Comment by spaze** [comment]
Different lexer setup results in the same crash, although the nodes are still different, but honestly I have no idea what I'm doing here, but I understand why I see the crash now, the `*TokenPos` seem relevant for the Visitor and it's not cool when they're missing.
```
	public function testVisitor(string $codeBefore, string $codeAfter): void
	{
		$lexer = new Emulative([
			'usedAttributes' => [
				'comments',
				'startLine', 'endLine',
//				'startTokenPos', 'endTokenPos',
			],
		]);
```

**Comment by ondrejmirtes** [comment]
Alright, we tried, it's a mystery 😊 Thank you!

**Comment by ondrejmirtes** [comment]
Tagged as 1.0.1.

File: src/Php/PhpPrinterIndentationDetectorVisitor.php [hunk 1]
```diff
@@ -39,7 +39,7 @@ public function enterNode(Node $node)
 			return null;
 		}
 
-		if (count($node->stmts) === 0) {
+		if ($node->stmts === null || count($node->stmts) === 0) {
```
Comment (staabm) [inline src/Php/PhpPrinterIndentationDetectorVisitor.php:42]: I wonder why phpstan did not error regarding count(mixed)

Comment (ondrejmirtes) [reply src/Php/PhpPrinterIndentationDetectorVisitor.php:42]: I don't use level 9 (but maybe I should for this small library).

Comment (spaze) [reply src/Php/PhpPrinterIndentationDetectorVisitor.php:42]: I've tried running PHPStan on level 9 here, and the results were the same before and after the patch: total 5 errors, none related to this `count()`.

**Pull request merged**